runtime/drivers/mysql/sql_store.go

```diff
@@ -0,0 +1,212 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	sqldriver "database/sql/driver"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/mitchellh/mapstructure"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime/drivers"
+)
+
+// Query implements drivers.SQLStore
+func (c *connection) Query(ctx context.Context, props map[string]any) (drivers.RowIterator, error) {
+	srcProps, err := parseSourceProperties(props)
+	if err != nil {
+		return nil, err
+	}
+
+	var dsn string
+	if srcProps.DSN != "" { // get from src properties
+		dsn = srcProps.DSN
+	} else if url, ok := c.config["dsn"].(string); ok && url != "" { // get from driver configs
+		dsn = url
+	} else {
+		return nil, fmt.Errorf("the property 'dsn' is required for MySQL. Provide 'dsn' in the YAML properties or pass '--env connector.mysql.dsn=...' to 'rill start'")
+	}
+
+	conf, err := mysql.ParseDSN(dsn)
+	if err != nil {
+		return nil, err
+	}
+	conf.ParseTime = true // if set to false, time is scanned as an array rather than as time.Time
+
+	db, err := sql.Open("mysql", conf.FormatDSN())
+	if err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	// Validate DSN data:
+	err = db.Ping()
+	if err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	rows, err := db.QueryContext(ctx, srcProps.SQL)
+	if err != nil {
+		return nil, err
+	}
+
+	iter := &rowIterator{
+		db:   db,
+		rows: rows,
+	}
+
+	if err := iter.setSchema(ctx); err != nil {
+		iter.Close()
+		return nil, err
+	}
+	return iter, nil
+}
+
+// QueryAsFiles implements drivers.SQLStore
+func (c *connection) QueryAsFiles(ctx context.Context, props map[string]any, opt *drivers.QueryOption, p drivers.Progress) (drivers.FileIterator, error) {
+	return nil, drivers.ErrNotImplemented
+}
+
+type rowIterator struct {
+	db   *sql.DB
+	rows *sql.Rows
+
+	schema       *runtimev1.StructType
+	row          []sqldriver.Value
+	fieldMappers []mapper
+}
+
+// Close implements drivers.RowIterator.
+func (r *rowIterator) Close() error {
+	r.rows.Close()
+	r.db.Close()
+	return nil
+}
+
+// Next implements drivers.RowIterator.
+func (r *rowIterator) Next(ctx context.Context) ([]sqldriver.Value, error) {
+	if !r.rows.Next() {
+		err := r.rows.Err()
+		if err == nil {
+			return nil, drivers.ErrIteratorDone
+		}
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, fmt.Errorf("no results found for the query")
+		}
+		return nil, err
+	}
+
+	cts, err := r.rows.ColumnTypes()
+	if err != nil {
+		return nil, err
+	}
+
+	dests := make([]any, len(cts))
+	for i := range dests {
+		dests[i], err = r.fieldMappers[i].dest(cts[i].ScanType())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err := r.rows.Scan(dests...); err != nil {
+		return nil, err
+	}
+
+	for i := range r.schema.Fields {
+		v := dests[i]
+		if v == nil {
+			r.row[i] = nil
+			continue
+		}
+		if reflect.ValueOf(v).Kind() == reflect.Ptr {
```

Comment:
Could you please explain this if block? Some questions:
1. Wouldn't this always be a ptr since dest always returns pointer to some type ?
2. Under what conditions can it be a slice ?

Reply:
1. Yes
2. Binary columns

However, I completely modified this block